### PR TITLE
Migrate from poetry to uv

### PR DESCRIPTION
**Pull request by jorenham**
cc @KotlinIsland

### CODE REVIEW

File: .pre-commit-config.yaml [hunk 1]
```diff
@@ -68,13 +63,13 @@ repos:
...
         language: system
         types_or: [python, pyi]
 
       - id: basedpyright-verifytypes
         name: basedpyright --verifytypes
-        entry: poetry run basedpyright --ignoreexternal --verifytypes mainpy
+        entry: uv run basedpyright --ignoreexternal --verifytypes mainpy
         language: system
         always_run: true
         pass_filenames: false
```
Comment (KotlinIsland) [inline .pre-commit-config.yaml:75]: it would be nice to have a deploy job, but maybe that for another PR

**Review by KotlinIsland — APPROVED** [review]

File: pyproject.toml [hunk 2]
```diff
@@ -1,72 +1,67 @@
 [build-system]
-requires = ["poetry-core>=1.0.0"]
-build-backend = "poetry.core.masonry.api"
+requires = ["hatchling"]
+build-backend = "hatchling.build"
```
[the remark this answers is not in the document]
Comment (KotlinIsland) [reply pyproject.toml:3]: additionally, `uv` has `build` now

File: .pre-commit-config.yaml (hunk 1 above)
Comment (jorenham) [reply .pre-commit-config.yaml:75]: Yes, that's something that's missing from almost all of my repo's actually. So you think that building + publishing based on a tag-push is something that'll work?

File: pyproject.toml (hunk 2 above)
[the remark this answers is not in the document]
Comment (jorenham) [reply pyproject.toml:3]: > is this required for sdists?

No idea, just copied this over from some sketchy guide 🤷🏻. You think it isn't?

File: .pre-commit-config.yaml (hunk 1 above)
Comment (KotlinIsland) [reply .pre-commit-config.yaml:75]: can do a strategy like https://github.com/KotlinIsland/basedmypy/issues/167

basically:
- create draft release on every merge to `main`
- on publish of draft release, trigger publish job

File: pyproject.toml [hunk 3]
```diff
@@ -1,72 +1,67 @@
...
-    "Programming Language :: Python :: 3.13",
+  "Development Status :: 5 - Production/Stable",
+  "Operating System :: OS Independent",
+  "Typing :: Typed",
+  "Programming Language :: Python :: 3.8",
+  "Programming Language :: Python :: 3.9",
+  "Programming Language :: Python :: 3.10",
+  "Programming Language :: Python :: 3.11",
+  "Programming Language :: Python :: 3.12",
+  "Programming Language :: Python :: 3.13",
```
Comment (KotlinIsland) [inline pyproject.toml:24]: these were automatic with poetry, I wonder if it's supported in `uv`

File: .pre-commit-config.yaml (hunk 1 above)
Comment (KotlinIsland) [reply .pre-commit-config.yaml:75]: needs https://github.com/astral-sh/uv/issues/6278

File: .github/workflows/ci.yml [hunk 4]
```diff
@@ -20,29 +20,29 @@ jobs:
       - name: checkout
         uses: actions/checkout@v4
 
-      - name: install poetry
-        run: pipx install poetry
+      - name: setup uv
+        uses: eifinger/setup-uv@v1
```
Comment (KotlinIsland) [inline .github/workflows/ci.yml:24]: `astral-sh/setup-uv@v2`?

File: CONTRIBUTING.md [hunk 5]
```diff
@@ -176,19 +176,19 @@ which could serve as inspiration.
...
-It can help to use mainpy's lowest-supported Python version, so that you don't
-accidentally use those bleeding-edge Python features that you shouldn't, e.g.
+Ensure you have [uv](https://github.com/astral-sh/uv) installed.
+It can help to use mainpy's lowest-supported Python version (3.10 for
+development), so that you don't accidentally use those bleeding-edge Python
+features that you shouldn't, e.g.
 
 ```bash
-poetry env use python3.11
+uv python install 3.10
```
Comment (KotlinIsland) [inline CONTRIBUTING.md:185]: `uv sync --python 3.10`

**Review by KotlinIsland — APPROVED** [review]

File: pyproject.toml (hunk 2 above)
[the remark this answers is not in the document]
Comment (KotlinIsland) [reply pyproject.toml:3]: this is required for sdists https://docs.astral.sh/uv/concepts/projects/#creating-projects

File: .pre-commit-config.yaml (hunk 1 above)
Comment (jorenham) [reply .pre-commit-config.yaml:75]: there's an issue for this now: https://github.com/jorenham/mainpy/issues/45

File: .github/workflows/ci.yml [hunk 6]
```diff
@@ -20,29 +20,29 @@ jobs:
...
-        uses: actions/setup-python@v5
-        with:
-          python-version: 3.8
-          cache: poetry
+        run: uv python install 3.8
 
-      - name: install
-        run: poetry install
+      - name: install dependencies
+        run: uv sync --frozen --all-extras --dev
```
[the remark this answers is not in the document]
Comment (jorenham) [reply .github/workflows/ci.yml:33]: That's what `sync --frozen` also does:

> Display the requirements without locking the project.
> 
> If the lockfile is missing, uv will exit with an error.

File: pyproject.toml (hunk 3 above)
Comment (jorenham) [reply pyproject.toml:24]: UV's own [`pyproject.toml`](https://github.com/astral-sh/uv/blob/main/pyproject.toml) has them, so I suppose they should still be added manually

**Comment by jorenham** [comment]
I don't think we can use UV to test on python 3.13 yet 🤔 
https://github.com/astral-sh/uv/issues/6697

**Comment by jorenham** [comment]
I notices that `uv build` produces an sdist archive that's quite a bit larger than that of  `poetry build`:

```
-rw-r--r-- 1 joren joren 5,6K sep  2 15:27 mainpy-1.4.0-py3-none-any.whl
-rw-rw-r-- 1 joren joren 7,0K sep  2 15:27 mainpy-1.4.0.tar.gz
-rw-r--r-- 1 joren joren 5,6K sep  8 19:55 mainpy-1.4.1.dev0-py3-none-any.whl
-rw-r--r-- 1 joren joren  34K sep  8 19:55 mainpy-1.4.1.dev0.tar.gz
```

It appears that `uv build` also includes the irrelevant files in the sdist, such as `.github/` and `tests/`, so basically everything in the repo...

**Comment by jorenham** [comment]
Given the lack of `uv publish`, and the bugs in `uv build`, I think it's better to wait until those are resolved before swtiching from poetry to uv.

File: CONTRIBUTING.md [hunk 7]
```diff
@@ -176,19 +176,15 @@ which could serve as inspiration.
...
-```bash
-poetry env use python3.11
-```
-
-Now you can install the dev dependencies using
+You can install the lowest supported Python version with
 
 ```bash
-poetry install --sync
+uv sync --dev --python 3.10
```
Comment (KotlinIsland) [inline CONTRIBUTING.md:187]:
```suggestion
uv sync --python 3.10
```
https://docs.astral.sh/uv/reference/cli/#uv-sync

**Comment by KotlinIsland** [comment]
> I don't think we can use UV to test on python 3.13 yet 🤔 [astral-sh/uv#6697](https://github.com/astral-sh/uv/issues/6697)

> Alternatively, the official GitHub setup-python action can be used. This can be faster, because GitHub caches the Python versions alongside the runner.

https://docs.astral.sh/uv/guides/integration/github/#setting-up-python

**Comment by KotlinIsland** [comment]
> It appears that uv build also includes the irrelevant files in the sdist, such as `.github/` and `tests/`, so basically everything in the repo...

i wonder if this is regarding differing selection rules

**Comment by jorenham** [comment]
@KotlinIsland I'm sold on `uv`. You can merge this if you want

**Review by KotlinIsland — APPROVED** [review]

**Pull request merged**